pages/docs/reason-compiler/latest/lazy-values.mdx

```diff
@@ -0,0 +1,129 @@
+# Lazy Values
+
+A lazy value of represents a deferred computation which will automatically memoize the result on the first run, and then return the memoized result on any repeated execution.
+
+This is useful for e.g. defining functions and expressions for complex procedures that always return the same value, for example:
+
+- Doing expensive DOM traversals over the same tree over and over again
+- Doing file system operations on a static set of files that won't change 
+- Doing expensive requests to an API server that would always return the same data
+
+A lazy value has a type of `Lazy.t('a)`, where `'a` is the return value of the computation. All its functionality is encapsulated with the globally available `Lazy` module.
+
+## Creating a lazy value
+
+Lazy values are part of the language. You can either use the `lazy` keyword to create a lazy value from an expression...
+
+```re example
+// We only want getFiles to read the file system once,
+// so we wrap it in a lazy value
+let getFiles =
+  lazy({
+    Js.log("Reading dir");
+    Node.Fs.readdirSync("./pages");
+  });
+
+// On the first call, the computation takes place
+Lazy.force(getFiles)->Js.log;
+
+// The second call will just return the already calculated files
+Lazy.force(getFiles)->Js.log;
+```
+
+... or you can also wrap an existing function to make it lazy:
+
+```reason example
+let getFiles = () => { Node.Fs.readdirSync("./pages"); };
+
+// Note: getFiles() is not actually called within the lazy expression!
+let lazyGetFiles = lazy(getFiles());
+```
+
+Note that `getFiles()` inside the `lazy` expression is not evaluated right away (its type will be `Lazy.t(array(string))`! This is because `lazy(...)` is not a function and will be treated differently by the compiler.
+
+Alternatively we can use `Lazy.from_fun(getFiles)` to create the same lazy value.
```

Comment:
If we have an existing function of type `unit => something` then `Lazy.from_fun(theFun)` should be preferred, otherwise `lazy(theValue)`.